core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -45,27 +46,27 @@ private[log] case object LogCleaningPaused extends LogCleaningState
  *  While a partition is in the LogCleaningPaused state, it won't be scheduled for cleaning again, until cleaning is
  *  requested to be resumed.
  */
-private[log] class LogCleanerManager(val logDirs: Array[File], val logs: Pool[TopicPartition, Log]) extends Logging with KafkaMetricsGroup {
+private[log] class LogCleanerManager(val logDirs: ArrayBuffer[File], val logs: Pool[TopicPartition, Log]) extends Logging with KafkaMetricsGroup {
 
   import LogCleanerManager._
 
   override val loggerName = classOf[LogCleaner].getName
 
   // package-private for testing
   private[log] val offsetCheckpointFile = "cleaner-offset-checkpoint"
-  
+
   /* the offset checkpoints holding the last cleaned point for each log */
-  private val checkpoints = logDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, offsetCheckpointFile)))).toMap
+  @volatile private var checkpoints = logDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, offsetCheckpointFile)))).toMap
```

Comment:
If you have an external lock during updates in both cases (like in this example), then the volatile version has the advantage that you don't need a lock on reads. The disadvantage is that there is potentially more copying. In this particular case, you are using `filterKeys` which is just a view on the underlying map. That doesn't seem like a good idea. Aside from that, I have to look at the changes more closely to have a worthwhile opinion on what should be done here.